src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -690,38 +700,86 @@ private async Task SendRstStreamAsync(int streamId, Http2ProtocolErrorCode error
 
         private void WriteHeader(string name, string value)
         {
-            // TODO: ISSUE 31307: Use static table for known headers
-
             int bytesWritten;
             while (!HPackEncoder.EncodeHeader(name, value, _headerBuffer.AvailableSpan, out bytesWritten))
             {
-                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpan.Length + 1);
+                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpanLength + 1);
+            }
+
+            _headerBuffer.Commit(bytesWritten);
+        }
+
+        private void WriteHeaderName(ReadOnlySpan<byte> encodedName)
+        {
+            if (_headerBuffer.AvailableSpanLength < encodedName.Length)
+            {
+                _headerBuffer.EnsureAvailableSpace(encodedName.Length);
             }
+            encodedName.CopyTo(_headerBuffer.AvailableSpan);
+            _headerBuffer.Commit(encodedName.Length);
+        }
+
+        private void WriteHeader(ReadOnlySpan<byte> encodedName, string value)
+        {
+            // Copy the already encoded name
+            if (_headerBuffer.AvailableSpanLength < encodedName.Length)
+            {
+                _headerBuffer.EnsureAvailableSpace(encodedName.Length);
+            }
+            encodedName.CopyTo(_headerBuffer.AvailableSpan);
+            _headerBuffer.Commit(encodedName.Length);
 
+            // Encode and write the value
+            int bytesWritten;
+            while (!HPackEncoder.EncodeHeaderValue(value, _headerBuffer.AvailableSpan, out bytesWritten))
+            {
+                _headerBuffer.EnsureAvailableSpace(_headerBuffer.AvailableSpanLength + 1);
+            }
             _headerBuffer.Commit(bytesWritten);
         }
 
+        private void WriteHeader(ReadOnlySpan<byte> encodedHeader)
+        {
+            Debug.Assert(encodedHeader != null && encodedHeader.Length > 0);
```

Comment:
How can encodedHeader be null?

Also, should we add the Length assert to similar routines above?

Reply:
It can't be. Leftover from when I had it as a byte[].